frontend/components/data-table.js

```diff
@@ -0,0 +1,550 @@
+
+export default class DataTable extends Component {
+  constructor({ element, phones, columnConfig }) {
+    super({ element });
+
+    this._props = {
+      config: columnConfig,
+      defaultPhones: phones,
+      phonesLength: phones.length,
+      phones,
+    };
+
+    this._render();
+    this._init();
+  }
+
+  _init() {
+    const defaultPerPage = 5;
+    const { phonesLength } = this._props;
+
+    this._state = {
+      filterValue: '',
+      selectedFilter: 'all',
+      orderValue: 'up',
+      orderName: '',
+      mainCheckboxValue: false,
+      currentPage: 0,
+      totalItemsCount: this.getItemsCount(),
+      buttonsCount: Math.ceil(phonesLength / defaultPerPage),
+      itemsPerPage: defaultPerPage,
+      perPageOptions: [3, 5, 10, 15, 20],
+    };
+
+    let { defaultPhones } = this._props;
+
+    defaultPhones = this._addCheckFieldToItems(defaultPhones);
+    this.updateProps({ defaultPhones });
+
+    this._initFilter();
+    this._initPagination();
+    this._initPhonesCheckbox();
+    this._initFieldsEditing();
+
+    this._makeFiltering();
+  }
+
+  _initFilter() {
+    this._filterInput = this._getElement('find-input');
+
+    this.on('click', '[data-type="filter-selected"]', (event) => {
+      this._clearCheckButtonsStyle();
+
+      const btn = event.target.closest('[data-type="filter-selected"]');
+      btn.classList.add('header-button-active');
+      const selectedFilter = btn.dataset.value;
+
+      this._updateState({ selectedFilter });
+      this._resetCurrentPage();
+      this._makeFiltering();
+
+      const { itemsPerPage, buttonsCount } = this._state;
+      const { phonesLength } = this._props;
+
+      this._pagination.updateProps({
+        perPage: itemsPerPage,
+        phonesLength,
+        buttonsCount,
+      });
+    });
+
+    this.on('input', '[data-element="find-input"]', () => {
+      this.emit('input-enter');
+    });
+
+    this.on('click', '[data-sortable-key]', (event) => {
+      const element = event.target.closest('[data-sortable-key]');
+      const currOrderName = element.dataset.sortableKey;
+
+      let { orderName, orderValue } = this._state;
+
+      if (orderName === currOrderName) {
+        orderValue = orderValue === 'up' ? 'down' : 'up';
+      } else {
+        orderName = currOrderName;
+      }
+
+      this._updateState({ orderValue, orderName });
+      this.emit('order-enter');
+    });
+
+    this.subscribe(
+      'input-enter',
+      this._debounce(() => {
+        this._resetCurrentPage();
+        this._makeFiltering();
+        const {
+          currentPage,
+          itemsPerPage,
+          buttonsCount,
+        } = this._state;
+
+        const { phonesLength } = this._props;
+        this._pagination.updateProps({
+          currentPage,
+          itemsPerPage,
+          phonesLength,
+          buttonsCount,
+        });
+      }, 500),
+    );
+
+    this.subscribe('order-enter', this._makeFiltering.bind(this));
```

Comment:
subscribing to own events is a bad idea
it makes the internal logic more complicated
just call the method where you emitted an event